server/appServices.js

```diff
@@ -100,27 +97,86 @@ function updateAppsAndWatch(namespace, kubeclient) {
       }
     });
 
-    const androidVariantStream = kubeclient.apis[androidVariantCRD.spec.group].v1alpha1.watch
-      .namespace(namespace)
-      .androidvariants.getStream();
-    const androidVariantJsonStream = new JSONStream();
-    androidVariantStream.pipe(androidVariantJsonStream);
-    androidVariantJsonStream.on('data', event => {
-      if (event.object && event.object.status) {
-        updateAll(namespace, kubeclient);
-      }
-    });
+    watchAndroidVariantStream(namespace, kubeclient);
+    watchIosVariantStream(namespace, kubeclient);
+  });
+}
 
-    const iosVariantStream = kubeclient.apis[iosVariantCRD.spec.group].v1alpha1.watch
-      .namespace(namespace)
-      .iosvariants.getStream();
-    const iosVariantJsonStream = new JSONStream();
-    iosVariantStream.pipe(iosVariantJsonStream);
-    iosVariantJsonStream.on('data', event => {
-      if (event.object && event.object.status) {
-        updateAll(namespace, kubeclient);
-      }
-    });
+/**
+ * Watch for updates to the MobileClient custom resource.
+ * @param {String} namespace - The namespace to watch
+ * @param {*} kubeclient - kubernetes client
+ */
+function watchMobileClients(namespace, kubeclient) {
+  const appStream = kubeclient.apis[mobileClientCRD.spec.group].v1alpha1.watch
+    .namespaces(namespace)
+    .mobileclients.getStream();
+  const appJsonStream = new JSONStream();
+  appStream.pipe(appJsonStream);
+  appStream.on('data', event => {
+    if (event.type === 'ADDED') {
+      updateAll(namespace, kubeclient);
+    }
+  });
+
+  // === WORKAROUND ===
+  // The stream automatically closes when watching CRDs after an inconsistent amount of time.
+  // Listen for when the stream ends and reopen it.
+  // Ref: https://github.com/godaddy/kubernetes-client/issues/520
+  appStream.on('end', () => {
```

Comment:
actually, is there an `error` event that will be emitted, or does it always emit the `end` event?

Reply:
https://www.npmjs.com/package/JSONStream

There doesn't seem to be.